### DISCUSSION TITLE
Unnecessary api call on invalidating a provider

### DISCUSSION THREAD

**Discussion by yasiralijaved** [rrousselGit/riverpod#3548] [Q&A]
So i am using ProfileProvider to fetch the user profile from rest api.

```
@Riverpod(keepAlive: false)
class Profile extends _$Profile {
  @override
  Future<User> build() async {
    await load();
    return await future;
  }
  
  load() async {
       state = const AsyncValue.loading();
       // fetch user profile from http client
       .
       .
       state = AsyncValue.data(user);
  }
}
```

On logout when i try to invalidate the the profile provider, the build() of profile provider called thus calling the api again. This is because the value of profile provider is still being watched in HomeWidget until the HomeWidget gets destroyed and LoginWidget comes up.

> logout button press in HomeWidget

```
onPressed: () {
                  ref.invalidate(profileProvider);
                  ref.invalidate(projectListProvider);
                  ref.read(accessTokenProvider.notifier).logout();
                },
```

> profile watcher in HomeWidget

```
class _HomeViewState extends ConsumerState<_HomeView> {
  @override
  Widget build(BuildContext context) {
    final state = ref.watch(profileProvider);

    String name = state.when(
        data: (user) => "${user.firstName} ${user.lastName}",
        error: (error, stack) => "Error while fetching user profile",
        loading: () => "Loading...");
```

I know this can be avoided if there is no watcher available for this profile provider. But there is no way i can unsubscribe from profile provider in the HomeWidget.

I want to avoid the build() being called on logout. Help is required.

**Comment by snapsl** [comment]
The `onPressed()` handler looks weird. There should be no need to invalidate profileProvider and projectListProvider.
In general, the login state is usually handled in the router (e.g. auth => HomeScreen, noauth=>LoginScreen). Maybe have a look how other implement riverpod + auth + router.

**Reply by yasiralijaved** [reply]
The onPressed of logout button is invalidating the providers because i need these providers to go to their default state (e.g: null profile data), so that next user logs in, these providers fetch the fresh data. Otherwise these providers would give the older data (e.g the profile of previous user, which is dangerous).

**Reply by snapsl** [reply]
>  Otherwise these providers would give the older data (e.g the profile of previous user, which is dangerous).

I don't have the full context here, but this should not be the case. 
If you do not explicitly keep these providers alive, they should be autodisposed when switching to the login screen.

**Reply by yasiralijaved** [reply]
I don't know for some reason the provider was autodisposing with  @Riverpod(keepAlive: true). That's i am trying to invalidate it on logout.

**Comment by charlescyt** [comment]
```dart
ref.invalidate(profileProvider);
ref.invalidate(projectListProvider);
ref.read(accessTokenProvider.notifier).logout();
```
This should be done in a declarative way. Your `profileProvider` & `projectListProvider` depends on the login state, and therefore should be watching a provider (`accessTokenProvider`) that represent the login state.

When the login state changes, those providers will be recomputed.
